### ISSUE TITLE
SequentialLR have a bug and why it use `step(epoch)`

### ISSUE THREAD

**Issue by HaoRan-hash** [pytorch/pytorch#67586]
I used PyTorch 1.10 immediately after it came out. I find `torch.optim.lr_scheduler.SequentialLR` have a bug.
The code snippet is below.
```
import torch
from torch import nn
import numpy as np
from torch.optim.lr_scheduler import SequentialLR, LinearLR, ExponentialLR
import matplotlib.pyplot as plt


class MyModel(nn.Module):
    def __init__(self):
        super(MyModel, self).__init__()
        self.layers = nn.Sequential(nn.Conv2d(3, 5, 3, 1, 0, bias=False),
                                    nn.BatchNorm2d(5),
                                    nn.ReLU())
    
    def forward(self, x):
        y = self.layers(x)
        return y


my_model = MyModel()

optimizer = torch.optim.SGD(my_model.parameters(), lr=0.1)
# scheduler1 = ExponentialLR(optimizer, gamma=0.5)
scheduler1 = LinearLR(optimizer, total_iters=5)
scheduler2 = ExponentialLR(optimizer, gamma=0.9)
scheduler = SequentialLR(optimizer, schedulers=[scheduler1, scheduler2], milestones=[5])

epochs = 50
lrs = []
for i in range(epochs):
    temp = optimizer.state_dict()
    lrs.append(temp['param_groups'][0]['lr'])
    scheduler.step()

x = list(range(epochs))
plt.plot(x, lrs)
plt.show()
```
When `scheduler1` use `LinearLR`, the training strategy is warmup. It's correct, and the `lrs` plot is below.
![Figure_1](https://user-images.githubusercontent.com/54461374/139524759-e1e4f322-a57f-44f5-8609-c9582aab147f.png)
But when `scheduler1` use `ExponentialLR`, the change in lr is not what I expected. The `lrs` plot is below.
![Figure_1](https://user-images.githubusercontent.com/54461374/139524851-6642b9c4-c2b6-45e8-8ab7-f1dd358536f9.png)
As you can see from the figure, the lr for `scheduler2` starts with the initial value again. I think it's not the literal mean of SequentialLR. I think `scheduler2` initial lr should be equal to` scheduler1` final lr.

Then I saw the source code. You use `lr_scheduler.step(0)`->`_get_closed_form_lr()` to do the SequentialLR. `_get_closed_form_lr()` calculate lr based on `lr_scheduler.base_lrs`, `lr_scheduler.base_lrs` is the initial lr, so it has the above question.
Also, a warning is generated at runtime because of step(0).
```
UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use scheduler.step() to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.
```

To summarize:

1. I think you should fix SequentialLR so that the initial LR of each phase scheduler is equal to the final LR of the previous phase scheduler.
2. I wonder why you use step(epoch), isn't it going to be deprecated?

**Comment by MichalLebeda** [comment]
Still does not work, any tips on how to chain schedulers without jumps?

**Comment by DavidNemeskey** [comment]
I ran into this as well. The current behavior is very counter-intuitive; I would have expected the later schedulers to continue where the previous one left off.

I "fixed" it by copying the class to my module and changing the `base_lrs` of the next scheduler to `self._last_lr` (a naming inconsistency there, BTW.) when a switch occurs. It seems to work, but I am not sure it's entirely correct.

**Comment by aneeshbhattacharya** [comment]
I fixed it by changing the definition in torch.optim.lr_scheduler file. 
First, in the scheduler logic add a function to set_base_lr (In my example I used 2 x CosineAnnealingLR Schedulers).
Secondly, in the SequentialLR, call the  set_base_lr function when the scheduler takes step(0).
Then it should look like this:
![lr_scheduler](https://github.com/pytorch/pytorch/assets/68210639/0ef7b749-4fd0-4681-803c-7cce9fe8aa86)
